flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/ConstraintEnforcerTestPrograms.java

```diff
@@ -0,0 +1,823 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.nodes.exec.stream;
+
+import org.apache.flink.table.api.TableRuntimeException;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.api.config.ExecutionConfigOptions.NestedEnforcer;
+import org.apache.flink.table.api.config.ExecutionConfigOptions.NotNullEnforcer;
+import org.apache.flink.table.api.config.ExecutionConfigOptions.TypeLengthEnforcer;
+import org.apache.flink.table.runtime.operators.sink.constraint.ConstraintEnforcer;
+import org.apache.flink.table.test.program.SinkTestStep;
+import org.apache.flink.table.test.program.SourceTestStep;
+import org.apache.flink.table.test.program.TableTestProgram;
+import org.apache.flink.types.Row;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.table.api.config.ExecutionConfigOptions.TABLE_EXEC_SINK_NESTED_CONSTRAINT_ENFORCER;
+import static org.apache.flink.table.api.config.ExecutionConfigOptions.TABLE_EXEC_SINK_NOT_NULL_ENFORCER;
+import static org.apache.flink.table.api.config.ExecutionConfigOptions.TABLE_EXEC_SINK_TYPE_LENGTH_ENFORCER;
+
+/** Tests for {@link ConstraintEnforcer}. */
+public class ConstraintEnforcerTestPrograms {
+
+    public static final String SCHEMA_NOT_NULL =
+            "a ROW<"
+                    + "id BIGINT NOT NULL, "
+                    + "ab ROW<"
+                    + "     aba BIGINT NOT NULL,"
+                    // extra column to make it easier to create a row with null as `aba`, without
+                    // the
+                    // column Row.of(null) can not differentiate if the vararg is null or the column
+                    // is null
+                    + "     ignore BIGINT"
+                    + "> NOT NULL> NOT NULL";
+    static final TableTestProgram NOT_NULL_DROP_NESTED_ROWS =
```

Comment:
Could we add one test for a structured type as well? This is can be done by implementing a ScalarFunction that returns a POJO. It will be made easier soon.

Reply:
Good that you asked, because I totally forgot about STRUCTURED_TYPE.

Unfortunately, I am not sure we can easily add a test for it, especially in SemanticTests infra. In order for a `ConstraintEnforcer` to kick in a given type need to be consumed by the sink. If I am not mistaken there is no way to define a `STRUCTURED_TYPE` in DDL and thus there is no way to make the sink consume a STRUCTURED_TYPE.

Happy to hear if you have an idea how to do that.